File: python/src/baselode/drill/intervals.py
# Copyright (C) 2026 Darkmine Pty Ltd
# SPDX-License-Identifier: GPL-3.0-or-later

"""From-to interval algebra primitives.

Pure, dependency-light operations on drillhole interval tables.  All
functions accept a pandas ``DataFrame`` keyed by ``HOLE_ID`` with from-/to-
depth columns, and return either a derived ``Series`` (lengths, midpoints)
or a new ``DataFrame``.  Functions never mutate inputs.

The module sits underneath :mod:`baselode.drill.composite` and
:mod:`baselode.drill.validate`, providing the boundary-arithmetic
primitives those higher-level operations consume.
"""

import numpy as np
import pandas as pd

from baselode.datamodel import FROM, HOLE_ID, TO


def interval_length(df, from_col=FROM, to_col=TO):
    """Length of each interval row.

    Parameters
    ----------
    df : pd.DataFrame
        Interval table with from-/to-depth columns.
    from_col, to_col : str
        Column names for the from- and to-depths.

    Returns
    -------
    pd.Series
        Length per row (``to - from``), indexed like *df*.
    """
    return df[to_col] - df[from_col]


def from_to_midpoints(df, from_col=FROM, to_col=TO):
    """Midpoint depth of each interval row.

    Parameters
    ----------
    df : pd.DataFrame
        Interval table with from-/to-depth columns.
    from_col, to_col : str
        Column names for the from- and to-depths.

    Returns
    -------
    pd.Series
        ``(from + to) / 2`` per row, indexed like *df*.
    """
    return (df[from_col] + df[to_col]) / 2.0


def detect_gaps(df, from_col=FROM, to_col=TO, hole_col=HOLE_ID, min_gap=0.0):
    """Uncovered downhole ranges between consecutive intervals, per hole.

    Sorts intervals by from-depth within each hole.  A gap is reported
    whenever the next from-depth exceeds the running maximum to-depth by
    more than *min_gap*.  Surface-to-first-interval and last-interval-to-EOH
    gaps are not reported (no anchor to compare against).

    Parameters
    ----------
    df : pd.DataFrame
        Interval table.
    from_col, to_col : str
        From-/to-depth columns.
    hole_col : str
        Hole identifier column.
    min_gap : float
        Minimum gap length to report.  Defaults to ``0.0`` (any positive
        gap reported).

    Returns
    -------
    pd.DataFrame
        One row per gap with columns ``hole_id``, ``from``, ``to``,
        ``length``.  Empty (with the same columns) when no gaps exist.
    """
    out_cols = [hole_col, from_col, to_col, "length"]
    if df.empty:
        return pd.DataFrame(columns=out_cols)

    gaps = []
    for hole_id, group in df.groupby(hole_col):
        group_sorted = group.sort_values(from_col)
        prev_to = None
        for _, row in group_sorted.iterrows():
            from_depth = float(row[from_col])
            to_depth = float(row[to_col])
            if prev_to is not None and from_depth - prev_to > min_gap:
                gaps.append({
                    hole_col: hole_id,
                    from_col: prev_to,
                    to_col: from_depth,
                    "length": from_depth - prev_to,
                })
            if prev_to is None or to_depth > prev_to:
                prev_to = to_depth

    if not gaps:
        return pd.DataFrame(columns=out_cols)
    return pd.DataFrame(gaps, columns=out_cols)


def detect_overlaps(df, from_col=FROM, to_col=TO, hole_col=HOLE_ID):
    """Pairwise overlapping intervals, per hole.

    All ordered pairs of intervals within a hole whose downhole ranges
    intersect are reported.  The reported ``from``/``to`` columns give the
    intersection range; ``length`` is its width.

    Parameters
    ----------
    df : pd.DataFrame
        Interval table.
    from_col, to_col : str
        From-/to-depth columns.
    hole_col : str
        Hole identifier column.

    Returns
    -------
    pd.DataFrame
        One row per overlapping pair with columns ``hole_id``, ``from``,
        ``to``, ``length``, ``first_index``, ``second_index`` (positional
        indices into *df*).
    """
    out_cols = [hole_col, from_col, to_col, "length", "first_index", "second_index"]
    if df.empty:
        return pd.DataFrame(columns=out_cols)

    overlaps = []
    df_with_pos = df.reset_index(drop=True).copy()
    df_with_pos["_pos"] = np.arange(len(df_with_pos))
    for hole_id, group in df_with_pos.groupby(hole_col):
        group_sorted = group.sort_values(from_col).reset_index(drop=True)
        rows = group_sorted.to_dict("records")
        row_count = len(rows)
        for first_idx in range(row_count):
            first_from = float(rows[first_idx][from_col])
            first_to = float(rows[first_idx][to_col])
            for second_idx in range(first_idx + 1, row_count):
                second_from = float(rows[second_idx][from_col])
                if second_from >= first_to:
                    break
                second_to = float(rows[second_idx][to_col])
                overlap_from = max(first_from, second_from)
                overlap_to = min(first_to, second_to)
                if overlap_to > overlap_from:
                    overlaps.append({
                        hole_col: hole_id,
                        from_col: overlap_from,
                        to_col: overlap_to,
                        "length": overlap_to - overlap_from,
                        "first_index": int(rows[first_idx]["_pos"]),
                        "second_index": int(rows[second_idx]["_pos"]),
                    })

    if not overlaps:
        return pd.DataFrame(columns=out_cols)
    return pd.DataFrame(overlaps, columns=out_cols)


def split_at(df, depths, from_col=FROM, to_col=TO, hole_col=HOLE_ID):
    """Split intervals at boundary depths.

    For each row, any *depth* falling strictly inside ``(from, to)`` is
    introduced as a new boundary; the row is replaced by the resulting
    sub-intervals, inheriting all other columns from the parent row.

    Parameters
    ----------
    df : pd.DataFrame
        Interval table.
    depths : dict, pd.DataFrame, list, or float
        Boundaries to split at, one of:

        - dict ``{hole_id: [d1, d2, ...]}`` — per-hole depths
        - DataFrame with ``hole_id`` and ``depth`` columns
        - list/array of depths — applied to every hole
        - single float — applied to every hole
    from_col, to_col : str
        From-/to-depth columns.
    hole_col : str
        Hole identifier column.

    Returns
    -------
    pd.DataFrame
        New table with the same columns as *df*; sub-intervals replace any
        row that contained a boundary.  Row order follows the input.
    """
    if df.empty:
        return df.copy()

    depths_by_hole = _normalize_depths_arg(depths, hole_col)

    out_rows = []
    for _, row in df.iterrows():
        hole_id = row[hole_col]
        from_depth = float(row[from_col])
        to_depth = float(row[to_col])

        if hole_id in depths_by_hole:
            candidate = depths_by_hole[hole_id]
        else:
            candidate = depths_by_hole.get("__ALL__", [])

        inner = sorted({float(depth) for depth in candidate if from_depth < float(depth) < to_depth})
        if not inner:
            out_rows.append(row.to_dict())
            continue

        boundaries = [from_depth, *inner, to_depth]
        for seg_from, seg_to in zip(boundaries[:-1], boundaries[1:]):
            new_row = row.to_dict()
            new_row[from_col] = seg_from
            new_row[to_col] = seg_to
            out_rows.append(new_row)

    return pd.DataFrame(out_rows, columns=df.columns)


def clip(df, from_depth=None, to_depth=None, from_col=FROM, to_col=TO):
    """Clip intervals to a downhole depth window.

    Intervals entirely outside ``[from_depth, to_depth]`` are dropped.
    Intervals straddling a boundary have their ``from`` and/or ``to`` value
    pulled to the window edge; all other columns are preserved.

    Parameters
    ----------
    df : pd.DataFrame
        Interval table.
    from_depth : float, optional
        Lower depth bound.  ``None`` means no lower bound.
    to_depth : float, optional
        Upper depth bound.  ``None`` means no upper bound.
    from_col, to_col : str
        From-/to-depth columns.

    Returns
    -------
    pd.DataFrame
        Clipped table; index is reset.
    """
    if df.empty:
        return df.copy()

    out = df.copy()
    if from_depth is not None:
        out = out.loc[out[to_col] > from_depth].copy()
        out.loc[:, from_col] = np.maximum(out[from_col], from_depth)
    if to_depth is not None:
        out = out.loc[out[from_col] < to_depth].copy()
        out.loc[:, to_col] = np.minimum(out[to_col], to_depth)
    return out.reset_index(drop=True)


def merge_tables(tables, from_col=FROM, to_col=TO, hole_col=HOLE_ID):
    """Left-join interval tables onto a common from-to support.

    Per hole, builds a fine-grained support by collecting all unique
    boundary depths across every input table, then emits one row per
    sub-interval with the non-key columns from each input table.  Within a
    sub-interval the lookup uses the midpoint, so each output row carries
    at most one source value per table.

    The first table in *tables* anchors the support: only depth ranges
    covered by it appear in the output ("left" semantics).  Columns from
    later tables are prefixed by their table name to avoid collisions.

    Parameters
    ----------
    tables : dict[str, pd.DataFrame]
        Ordered mapping ``{name: interval_table}``; the first entry is the
        left table.  Interval tables share *hole_col*, *from_col*, *to_col*
        but may carry any additional columns.
    from_col, to_col : str
        From-/to-depth columns shared across tables.
    hole_col : str
        Hole identifier column shared across tables.

    Returns
    -------
    pd.DataFrame
        Merged table with columns ``hole_id``, ``from``, ``to`` followed by
        ``<table>_<col>`` for each non-key column in each input.
    """
    if not tables:
        return pd.DataFrame(columns=[hole_col, from_col, to_col])

    table_names = list(tables.keys())
    left_name = table_names[0]
    left = tables[left_name]
    if left.empty:
        return pd.DataFrame(columns=[hole_col, from_col, to_col])

    holes = left[hole_col].unique()

    out_rows = []
    for hole_id in holes:
        left_for_hole = left[left[hole_col] == hole_id]
        boundaries = set()
        for _, row in left_for_hole.iterrows():
            boundaries.add(float(row[from_col]))
            boundaries.add(float(row[to_col]))

        left_min = float(left_for_hole[from_col].min())
        left_max = float(left_for_hole[to_col].max())

        for other_name in table_names[1:]:
            other = tables[other_name]
            other_for_hole = other[other[hole_col] == hole_id]
            for _, row in other_for_hole.iterrows():
                other_from = float(row[from_col])
                other_to = float(row[to_col])
                if other_to < left_min or other_from > left_max:
                    continue
                if other_from > left_min:
                    boundaries.add(other_from)
                if other_to < left_max:
                    boundaries.add(other_to)

        ordered = sorted(boundaries)

        for seg_from, seg_to in zip(ordered[:-1], ordered[1:]):
            midpoint = (seg_from + seg_to) / 2.0
            left_row = _row_containing(left_for_hole, midpoint, from_col, to_col)
            if left_row is None:
                continue
            out_row = {hole_col: hole_id, from_col: seg_from, to_col: seg_to}
            for col in left.columns:
                if col in (hole_col, from_col, to_col):
                    continue
                out_row[f"{left_name}_{col}"] = left_row[col]
            for other_name in table_names[1:]:
                other = tables[other_name]
                other_for_hole = other[other[hole_col] == hole_id]
                other_row = _row_containing(other_for_hole, midpoint, from_col, to_col)
                for col in other.columns:
                    if col in (hole_col, from_col, to_col):
                        continue
                    out_row[f"{other_name}_{col}"] = other_row[col] if other_row is not None else np.nan
            out_rows.append(out_row)

    if not out_rows:
        return pd.DataFrame(columns=[hole_col, from_col, to_col])
    return pd.DataFrame(out_rows)


def _row_containing(df, depth, from_col, to_col):
    """Return the row in *df* whose interval contains *depth*, or ``None``."""
    if df.empty:
        return None
    matches = df[(df[from_col] <= depth) & (df[to_col] > depth)]
    if matches.empty:
        return None
    return matches.iloc[0]


def _normalize_depths_arg(depths, hole_col):
    """Coerce ``split_at`` *depths* arg into ``{hole_id: [...]}`` form.

    A list/array/scalar is broadcast to every hole via the special
    ``"__ALL__"`` key, which :func:`split_at` interprets as "apply to any
    hole id".
    """
    if isinstance(depths, dict):
        return {hole_id: list(values) for hole_id, values in depths.items()}
    if isinstance(depths, pd.DataFrame):
        out = {}
        for hole_id, group in depths.groupby(hole_col):
            out[hole_id] = list(group["depth"])
        return out
    if np.isscalar(depths):
        return {"__ALL__": [float(depths)]}
    return {"__ALL__": [float(depth) for depth in depths]}

File: javascript/packages/baselode/src/data/intervals.js
/*
 * Copyright (C) 2026 Darkmine Pty Ltd
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

/**
 * From-to interval algebra primitives.
 *
 * Pure functions on arrays of interval rows.  All rows are plain objects
 * keyed by `HOLE_ID`, `FROM`, `TO` plus any additional value columns.
 * Functions never mutate inputs.
 */

import { HOLE_ID, FROM, TO } from './datamodel.js';

function groupByHole(rows, holeCol) {
  const grouped = new Map();
  for (const row of rows) {
    const hole = row[holeCol];
    if (hole == null) continue;
    if (!grouped.has(hole)) grouped.set(hole, []);
    grouped.get(hole).push(row);
  }
  return grouped;
}

/**
 * Length of each interval row (`to - from`).
 *
 * @returns {number[]}
 */
export function intervalLength(rows, { fromCol = FROM, toCol = TO } = {}) {
  return rows.map((row) => Number(row[toCol]) - Number(row[fromCol]));
}

/**
 * Midpoint depth of each interval row.
 *
 * @returns {number[]}
 */
export function fromToMidpoints(rows, { fromCol = FROM, toCol = TO } = {}) {
  return rows.map((row) => (Number(row[fromCol]) + Number(row[toCol])) / 2);
}

/**
 * Uncovered downhole ranges between consecutive intervals, per hole.
 *
 * @param {number} [options.minGap=0] - Minimum gap length to report.
 * @returns {Array<Object>} `{hole_id, from, to, length}` per gap.
 */
export function detectGaps(
  rows,
  { fromCol = FROM, toCol = TO, holeCol = HOLE_ID, minGap = 0 } = {},
) {
  if (!rows || !rows.length) return [];

  const result = [];
  for (const [hole, group] of groupByHole(rows, holeCol)) {
    const sorted = [...group].sort(
      (left, right) => Number(left[fromCol]) - Number(right[fromCol]),
    );
    let prevTo = null;
    for (const row of sorted) {
      const fromDepth = Number(row[fromCol]);
      const toDepth = Number(row[toCol]);
      if (prevTo !== null && fromDepth - prevTo > minGap) {
        result.push({
          [holeCol]: hole,
          [fromCol]: prevTo,
          [toCol]: fromDepth,
          length: fromDepth - prevTo,
        });
      }
      if (prevTo === null || toDepth > prevTo) prevTo = toDepth;
    }
  }
  return result;
}

/**
 * Pairwise overlapping intervals, per hole.
 *
 * Indices are positional into the input `rows` array.
 *
 * @returns {Array<Object>} `{hole_id, from, to, length, first_index, second_index}`.
 */
export function detectOverlaps(
  rows,
  { fromCol = FROM, toCol = TO, holeCol = HOLE_ID } = {},
) {
  if (!rows || !rows.length) return [];

  const indexed = rows.map((row, originalIndex) => ({ row, originalIndex }));
  const byHole = new Map();
  for (const entry of indexed) {
    const hole = entry.row[holeCol];
    if (hole == null) continue;
    if (!byHole.has(hole)) byHole.set(hole, []);
    byHole.get(hole).push(entry);
  }

  const result = [];
  for (const [hole, group] of byHole) {
    const sorted = [...group].sort(
      (left, right) => Number(left.row[fromCol]) - Number(right.row[fromCol]),
    );
    for (let firstIdx = 0; firstIdx < sorted.length; firstIdx += 1) {
      const firstFrom = Number(sorted[firstIdx].row[fromCol]);
      const firstTo = Number(sorted[firstIdx].row[toCol]);
      for (let secondIdx = firstIdx + 1; secondIdx < sorted.length; secondIdx += 1) {
        const secondFrom = Number(sorted[secondIdx].row[fromCol]);
        if (secondFrom >= firstTo) break;
        const secondTo = Number(sorted[secondIdx].row[toCol]);
        const overlapFrom = Math.max(firstFrom, secondFrom);
        const overlapTo = Math.min(firstTo, secondTo);
        if (overlapTo > overlapFrom) {
          result.push({
            [holeCol]: hole,
            [fromCol]: overlapFrom,
            [toCol]: overlapTo,
            length: overlapTo - overlapFrom,
            first_index: sorted[firstIdx].originalIndex,
            second_index: sorted[secondIdx].originalIndex,
          });
        }
      }
    }
  }
  return result;
}

function normalizeDepthsArg(depths, holeCol) {
  if (depths == null) return { byHole: new Map(), all: [] };
  if (Array.isArray(depths)) {
    return { byHole: new Map(), all: depths.map(Number) };
  }
  if (typeof depths === 'number') {
    return { byHole: new Map(), all: [Number(depths)] };
  }
  if (depths && typeof depths === 'object') {
    if (depths[holeCol] != null && depths.depth != null) {
      const byHole = new Map();
      byHole.set(depths[holeCol], [Number(depths.depth)]);
      return { byHole, all: [] };
    }
    const byHole = new Map();
    for (const [hole, value] of Object.entries(depths)) {
      const list = Array.isArray(value) ? value : [value];
      byHole.set(hole, list.map(Number));
    }
    return { byHole, all: [] };
  }
  return { byHole: new Map(), all: [] };
}

/**
 * Split intervals at boundary depths.
 *
 * Any depth strictly inside `(from, to)` is introduced as a new boundary;
 * the row is replaced by the resulting sub-intervals.  All other columns
 * are inherited unchanged.
 *
 * @param {(number|number[]|Object)} depths - One of: a scalar applied to
 *   every hole; an array applied to every hole; or
 *   `{hole_id: [d1, d2, ...]}`.
 * @returns {Array<Object>}
 */
export function splitAt(
  rows,
  depths,
  { fromCol = FROM, toCol = TO, holeCol = HOLE_ID } = {},
) {
  if (!rows || !rows.length) return [];

  const { byHole, all } = normalizeDepthsArg(depths, holeCol);

  const out = [];
  for (const row of rows) {
    const hole = row[holeCol];
    const fromDepth = Number(row[fromCol]);
    const toDepth = Number(row[toCol]);
    const candidate = byHole.has(hole) ? byHole.get(hole) : all;
    const inner = [...new Set(
      candidate
        .map(Number)
        .filter((depth) => depth > fromDepth && depth < toDepth),
    )].sort((left, right) => left - right);
    if (!inner.length) {
      out.push({ ...row });
      continue;
    }
    const boundaries = [fromDepth, ...inner, toDepth];
    for (let boundaryIdx = 0; boundaryIdx < boundaries.length - 1; boundaryIdx += 1) {
      out.push({
        ...row,
        [fromCol]: boundaries[boundaryIdx],
        [toCol]: boundaries[boundaryIdx + 1],
      });
    }
  }
  return out;
}

/**
 * Clip intervals to a downhole depth window.
 *
 * Intervals entirely outside the window are dropped; straddling intervals
 * have their from/to pulled to the boundary.  All other columns preserved.
 *
 * @returns {Array<Object>}
 */
export function clip(rows, fromDepth, toDepth, { fromCol = FROM, toCol = TO } = {}) {
  if (!rows || !rows.length) return [];

  const out = [];
  for (const row of rows) {
    let segFrom = Number(row[fromCol]);
    let segTo = Number(row[toCol]);
    if (fromDepth != null) {
      if (segTo <= fromDepth) continue;
      if (segFrom < fromDepth) segFrom = fromDepth;
    }
    if (toDepth != null) {
      if (segFrom >= toDepth) continue;
      if (segTo > toDepth) segTo = toDepth;
    }
    out.push({ ...row, [fromCol]: segFrom, [toCol]: segTo });
  }
  return out;
}

function rowContaining(group, depth, fromCol, toCol) {
  for (const row of group) {
    if (Number(row[fromCol]) <= depth && Number(row[toCol]) > depth) return row;
  }
  return null;
}

/**
 * Left-join interval tables onto a common from-to support.
 *
 * Per hole, the first ("left") table anchors the support; boundary depths
 * from later tables that fall inside the left's coverage are introduced as
 * new sub-interval boundaries.  Each output row carries `<name>_<col>`
 * values looked up at the sub-interval midpoint.  Where a later table has
 * no row covering the midpoint, the column is `null`.
 *
 * @param {Object<string, Array<Object>>} tables - Ordered object literal:
 *   the first key is the left table.
 * @returns {Array<Object>} Rows with `hole_id`, `from`, `to`, and
 *   prefixed columns from every input table.
 */
export function mergeTables(
  tables,
  { fromCol = FROM, toCol = TO, holeCol = HOLE_ID } = {},
) {
  if (!tables) return [];

  const tableNames = Object.keys(tables);
  if (!tableNames.length) return [];

  const leftName = tableNames[0];
  const left = tables[leftName] || [];
  if (!left.length) return [];

  const leftByHole = groupByHole(left, holeCol);
  const otherByHole = {};
  for (const otherName of tableNames.slice(1)) {
    otherByHole[otherName] = groupByHole(tables[otherName] || [], holeCol);
  }

  const out = [];
  for (const [hole, leftRowsForHole] of leftByHole) {
    const boundaries = new Set();
    let leftMin = Infinity;
    let leftMax = -Infinity;
    for (const row of leftRowsForHole) {
      const fromDepth = Number(row[fromCol]);
      const toDepth = Number(row[toCol]);
      boundaries.add(fromDepth);
      boundaries.add(toDepth);
      if (fromDepth < leftMin) leftMin = fromDepth;
      if (toDepth > leftMax) leftMax = toDepth;
    }
    for (const otherName of tableNames.slice(1)) {
      const otherRowsForHole = otherByHole[otherName].get(hole) || [];
      for (const row of otherRowsForHole) {
        const otherFrom = Number(row[fromCol]);
        const otherTo = Number(row[toCol]);
        if (otherTo < leftMin || otherFrom > leftMax) continue;
        if (otherFrom > leftMin) boundaries.add(otherFrom);
        if (otherTo < leftMax) boundaries.add(otherTo);
      }
    }
    const ordered = [...boundaries].sort((first, second) => first - second);

    for (let boundaryIdx = 0; boundaryIdx < ordered.length - 1; boundaryIdx += 1) {
      const segFrom = ordered[boundaryIdx];
      const segTo = ordered[boundaryIdx + 1];
      const midpoint = (segFrom + segTo) / 2;
      const leftRow = rowContaining(leftRowsForHole, midpoint, fromCol, toCol);
      if (!leftRow) continue;
      const outRow = { [holeCol]: hole, [fromCol]: segFrom, [toCol]: segTo };
      for (const col of Object.keys(leftRow)) {
        if (col === holeCol || col === fromCol || col === toCol) continue;
        outRow[`${leftName}_${col}`] = leftRow[col];
      }
      for (const otherName of tableNames.slice(1)) {
        const otherRowsForHole = otherByHole[otherName].get(hole) || [];
        const otherRow = rowContaining(otherRowsForHole, midpoint, fromCol, toCol);
        const sample = (tables[otherName] && tables[otherName][0]) || {};
        const otherCols = Object.keys(sample).filter(
          (col) => col !== holeCol && col !== fromCol && col !== toCol,
        );
        for (const col of otherCols) {
          outRow[`${otherName}_${col}`] = otherRow ? otherRow[col] : null;
        }
      }
      out.push(outRow);
    }
  }
  return out;
}
